langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java

```diff
@@ -1,542 +1,592 @@
 
-    /**
-     * Adds multiple embeddings and their corresponding contents that have been
-     * embedded to the store.
-     *
-     * @param embeddings
-     *     A list of embeddings to be added to the store.
-     * @param embedded
-     *     A list of original contents that were embedded.
-     * @return A list of auto-generated IDs associated with the added embeddings.
-     */
-    @Override
-    public List<String> addAll(List<Embedding> embeddings,
-        List<TextSegment> embedded) {
-        List<String> ids = createIds(embeddings);
-        addAllInternal(ids, embeddings, embedded);
-        return ids;
+        if (embeddings.size() != embedded.size()) {
+            throw new IllegalArgumentException("embeddings.size() " + embeddings.size()
+                    + " is not equal to embedded.size() " + embedded.size());
+        }
+
+        String[] ids = new String[embeddings.size()];
+
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement insert = connection.prepareStatement(
+                     "INSERT INTO " + tableName + "(id, embedding, content, metadata) VALUES (?, ?, ?, ?)")
+        ) {
+
+            for (int i = 0; i < embeddings.size(); i++) {
+                String id = ids[i] = randomUUID();
+                Embedding embedding = ensureIndexNotNull(embeddings, i, "embeddings");
+                TextSegment textSegment = ensureIndexNotNull(embedded, i, "embedded");
+
+                insert.setString(1, id);
+                insert.setObject(2, embedding.vector(), OracleType.VECTOR_FLOAT32);
+                insert.setObject(3, textSegment.text());
+                insert.setObject(4, getOsonFromMetadata(textSegment.metadata()));
+                insert.addBatch();
+            }
+            insert.executeBatch();
+
+            return Arrays.asList(ids);
+        }
+        catch (SQLException sqlException) {
+            throw uncheckSQLException(sqlException);
+        }
     }
 
-    /**
-     * Removes a single embedding from the store by ID.
-     *
-     * @param id The unique ID of the embedding to be removed.
-     */
     @Override
-    public void remove(String id) {
-        String deleteQuery = String.format("delete from %s where id = ?", table);
-        try (Connection connection = dataSource.getConnection(); PreparedStatement stmt = connection.prepareStatement(deleteQuery)) {
-            stmt.setString(1, id);
-            stmt.executeUpdate();
-        } catch (SQLException e) {
-            throw new RuntimeException("Failed to delete embedding " + id, e);
+    public void add(String id, Embedding embedding) {
+        ensureNotNull(id, "id");
+        ensureNotNull(embedding, "embedding");
+
+        // A MERGE command allows the user to update an existing embedding. The JavaDoc does not specify what this
+        // method should do in this case, but updating an existing row is consistent with other EmbeddingStore
+        // implementations.
```

Comment:
The idea is that the user-provided ID might be one of an existing table row. In this case, we would update the row, rather than insert a new row.

However, I'm not sure if that's actually a valid use case or not. Maybe the correct thing to do here is throw an error if the ID already exists in the table? I could not find a clear specification on what is correct.

All other add methods do not accept IDs from the user; The embedding store generates the IDs. For these methods, we use INSERT only because we know the IDs newly generated and don't exist in the table.

Does this make sense?

Reply:
The javadoc in the EmbeddingStore interface is not clear about the behavior. I guess updating makes sense. Thank you;

Reply:
Agreed!